ff/serving/serving/serving.go

```diff
@@ -0,0 +1,166 @@
+	logger.Info("Serving training data")
+	iter, err := serv.getTrainingSetIterator(name, variant)
+	if err != nil {
+		logger.Errorw("Failed to get training set iterator", "Error", err)
+		featureObserver.SetError()
+		return err
+	}
+	for iter.Next() {
+		sRow, err := serializedRow(iter.Features(), iter.Label())
+		if err != nil {
+			return err
+		}
+		if err := stream.Send(sRow); err != nil {
+			logger.Errorw("Failed to write to stream", "Error", err)
+			featureObserver.SetError()
+			return err
+		}
+		featureObserver.ServeRow()
+	}
+	if err := iter.Err(); err != nil {
+		logger.Errorw("Dataset error", "Error", err)
+		featureObserver.SetError()
+		return err
+	}
+	return nil
+}
+
+func (serv *FeatureServer) getTrainingSetIterator(name, variant string) (provider.TrainingSetIterator, error) {
+	ctx := context.TODO()
+	ts, err := serv.Metadata.GetTrainingSetVariant(ctx, metadata.NameVariant{name, variant})
+	if err != nil {
+		return nil, err
+	}
+	providerEntry, err := ts.FetchProvider(serv.Metadata, ctx)
+	if err != nil {
+		return nil, err
+	}
+	p, err := provider.Get(provider.Type(providerEntry.Type()), providerEntry.SerializedConfig())
+	if err != nil {
+		return nil, err
+	}
+	store, err := p.AsOfflineStore()
+	if err != nil {
+		// This means that the provider of the training set isn't an offline store.
+		// That shouldn't be possible.
+		return nil, err
+	}
+	return store.GetTrainingSet(provider.ResourceID{Name: name, Variant: variant})
+}
+
+func (serv *FeatureServer) FeatureServe(ctx context.Context, req *pb.FeatureServeRequest) (*pb.FeatureRow, error) {
+	features := req.GetFeatures()
+	entities := req.GetEntities()
+	entityMap := make(map[string]string)
+	for _, entity := range entities {
+		entityMap[entity.GetName()] = entity.GetValue()
+	}
+	vals := make([]*pb.Value, len(features))
+	for i, feature := range req.GetFeatures() {
+		name, variant := feature.GetName(), feature.GetVersion()
+		serv.Logger.Infow("Serving feature", "Name", name, "Variant", variant)
+		val, err := serv.getFeatureValue(ctx, name, variant, entityMap)
+		if err != nil {
+			return nil, err
+		}
+		vals[i] = val
+	}
+	return &pb.FeatureRow{
+		Values: vals,
+	}, nil
+}
+
+func (serv *FeatureServer) getFeatureValue(ctx context.Context, name, variant string, entityMap map[string]string) (*pb.Value, error) {
+	obs := serv.Metrics.BeginObservingOnlineServe(name, variant)
+	defer obs.Finish()
+	logger := serv.Logger.With("Name", name, "Variant", variant)
+	logger.Debug("Getting metadata")
+	meta, err := serv.Metadata.GetFeatureVariant(ctx, metadata.NameVariant{name, variant})
+	if err != nil {
+		logger.Errorw("metadata lookup failed", "Err", err)
+		obs.SetError()
+		return nil, err
+	}
+	entity, has := entityMap[meta.Entity()]
+	if !has {
+		logger.Errorw("Entity not found", "Entity", meta.Entity())
+		obs.SetError()
+		return nil, fmt.Errorf("No value for entity %s", meta.Entity())
+	}
+	providerEntry, err := meta.FetchProvider(serv.Metadata, ctx)
```

Comment:
maybe not necessary now but for the process of feature serving it would be good to do a simple in-memory cache of the providers/stores/tables if it's ever a use case to make really big feature requests with lots of entities/features